commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+ *    are easier from the user perspective. Even though it's a pull-based API,
+ *    it can be asynchronous underneath since the user can decide to work on
+ *    the traversable or iterator at a later stage.</li>
+ * </ul>
+ *
+ * @since 8.0
+ */
+public interface FunctionalMap<K, V> extends AutoCloseable {
+
+   /**
+    * Tweak functional map executions providing {@link Param} instances.
+    */
+   FunctionalMap<K, V> withParams(Param<?>... ps);
+
+   /**
+    * Functional map's name.
+    */
+   String getName();
+
+   /**
+    * Functional map's status.
+    */
+   Status getStatus();
+
+   /**
+    * Exposes read-only operations that can be executed against the functional map.
+    * The information that can be read per entry in the functional map is
+    * exposed by {@link ReadEntryView}.
+    *
+    * DESIGN RATIONALES:
+    * <ul>
+    *    <li>Why does it make sense to expose read-only operations?
+    *    Because read-only operations don't acquired locks, and hence all sorts
+    *    of optimizations can be carried out by the internal logic.
+    *    </li>
+    *    <li>Why no values() method? Having keys() makes sense since that way
+    *    we can have an observe all keys without having to bring values.
+    *    Having entries() makes sense since it allows you to observe on both
+    *    keys and values, but this is no extra cost to exposing just values
+    *    since keys are the main index and hence will always be available.
+    *    Hence, adding values() offers nothing extra to the API.
+    *    </li>
+    * </ul>
+    */
+   interface ReadOnlyMap<K, V> extends FunctionalMap<K, V> {
+      /**
+       * Tweak read-only functional map executions providing {@link Param} instances.
+       */
+      ReadOnlyMap<K, V> withParams(Param<?>... ps);
+
+      /**
+       * Evaluate a read-only function on the value associated with the key
+       * and return a {@link CompletableFuture} with the return type of the function.
+       * If the user is not sure if the key is present, {@link ReadEntryView#find()}
+       * can be used to find out for sure. Typically, function implementations
+       * would return value or {@link MetaParam} information from the cache
+       * entry in the functional map.
+       *
+       * By returning {@link CompletableFuture} instead of the function's
+       * return type directly, the method hints at the possibility that to
+       * executing the function might require remote data present in either
+       * a persistent store or a remote clustered node.
+       *
+       * This method can be used to implement read-only single-key based
+       * operations in {@link ConcurrentMap} and {@link javax.cache.Cache}
+       * such as:
+       *
+       * <ul>
+       *    <li>{@link ConcurrentMap#get(Object)}</li>
+       *    <li>{@link ConcurrentMap#containsKey(Object)}</li>
+       *    <li>{@link javax.cache.Cache#get(Object)}</li>
```

Comment:
I don't believe the javadoc link will work here since it is not in the classpath?